packages/semcom-upload-component/lib/components/sem-com-register.machine.ts

```diff
@@ -0,0 +1,363 @@
+
+  public type: SemComRegisterEvents.AUTHENTICATED = SemComRegisterEvents.AUTHENTICATED;
+  constructor(public session: Session) {}
+
+}
+
+export class StoreSelectedEvent implements EventObject {
+
+  public type: SemComRegisterEvents.STORE_SELECTED = SemComRegisterEvents.STORE_SELECTED;
+  constructor(public input: string) {}
+
+}
+
+export class HasPermissionEvent implements EventObject {
+
+  public type: SemComRegisterEvents.HAS_PERMISSION = SemComRegisterEvents.HAS_PERMISSION;
+  constructor(public url: string) {}
+
+}
+
+export class NoPermissionEvent implements EventObject {
+
+  public type: SemComRegisterEvents.NO_PERMISSION = SemComRegisterEvents.NO_PERMISSION;
+  constructor(public errorMessage: string) {}
+
+}
+
+export class BackToStoreSelectionEvent implements EventObject {
+
+  public type: SemComRegisterEvents.BACK_TO_STORE_SELECTION = SemComRegisterEvents.BACK_TO_STORE_SELECTION;
+
+}
+
+export class BackToUploadFormEvent implements EventObject {
+
+  public type: SemComRegisterEvents.BACK_TO_UPLOAD_FORM = SemComRegisterEvents.BACK_TO_UPLOAD_FORM;
+
+}
+
+export class UploadFormSubmittedEvent implements EventObject {
+
+  public type: SemComRegisterEvents.UPLOAD_FORM_SUBMITTED = SemComRegisterEvents.UPLOAD_FORM_SUBMITTED;
+  constructor (public uploadFormContext: UploadFormContext) {}
+
+}
+
+export class DataSavedEvent implements EventObject {
+
+  public type: SemComRegisterEvents.DATA_SAVED = SemComRegisterEvents.DATA_SAVED;
+
+}
+
+export class DataNotSavedEvent implements EventObject {
+
+  public type: SemComRegisterEvents.DATA_NOT_SAVED = SemComRegisterEvents.DATA_NOT_SAVED;
+  constructor(public errorMessage: string) {}
+
+}
+
+export type SemComRegisterEvent =
+| AuthenticatedEvent
+| StoreSelectedEvent
+| HasPermissionEvent
+| NoPermissionEvent
+| BackToStoreSelectionEvent
+| BackToUploadFormEvent
+| UploadFormSubmittedEvent
+| DataSavedEvent
+| DataNotSavedEvent;
+
+/* MACHINE */
+
+const handleStoreSelectedEvent = (event: StoreSelectedEvent) => {
+
+  const url = event.input;
+
+  if (!url) { return of(new NoPermissionEvent('Error retrieving the selected store. Please try again, or choose another store.')); }
+
+  return from(solidFetch(url, { method: 'HEAD' })).pipe(
+    switchMap((response) => {
+
+      if (response.headers.get('link') && response.headers.get('link').includes('<http://www.w3.org/ns/pim/space#Storage>; rel="type"')) {
+
+        const wacAllowHeader = response.headers.get('wac-allow');
+        const userRights = wacAllowHeader.replace(/.*user="([^"]*)".*/, '$1');
+
+        return userRights.includes('append') ? of(new HasPermissionEvent(url)) : of(new NoPermissionEvent('You do not have permission to edit this store. Please choose another store or log in with a user account that has access to this store.'));
+
+      }
+
+      return of(new NoPermissionEvent('The selected url did not lead to a valid store. Please choose another store.'));
+
+    }),
```

Comment:
You always return an `of(...)`, if possible, can we change this to `map(...)` and throw out the `of(...)`s?

Reply:
This one is actually on purpose. It must return an observable of the event, otherwise the invoke on line 300 errors:

```
Type '{ src: (context: SemComRegisterContext, event: StoreSelectedEvent) => Observable<NoPermissionEvent | Observable<NoPermissionEvent> | Observable<...>>; }' is not assignable to type 'SingleOrArray<StateMachine<any, any, any, { value: any; context: any; }> | InvokeConfig<SemComRegisterContext, SemComRegisterEvent>>'.
  Types of property 'src' are incompatible.
    Type '(context: SemComRegisterContext, event: StoreSelectedEvent) => Observable<NoPermissionEvent | Observable<NoPermissionEvent> | Observable<...>>' is not assignable to type 'string | StateMachine<any, any, any, { value: any; context: any; }> | InvokeSourceDefinition | InvokeCreator<SemComRegisterContext, SemComRegisterEvent, any>'.
      Type '(context: SemComRegisterContext, event: StoreSelectedEvent) => Observable<NoPermissionEvent | Observable<NoPermissionEvent> | Observable<...>>' is not assignable to type 'InvokeCreator<SemComRegisterContext, SemComRegisterEvent, any>'.
        Type 'Observable<NoPermissionEvent | Observable<NoPermissionEvent> | Observable<HasPermissionEvent>>' is not assignable to type 'StateMachine<any, any, any, { value: any; context: any; }> | Subscribable<EventObject> | Behavior<any, any> | PromiseLike<any> | InvokeCallback<...>'.
          Type 'Observable<NoPermissionEvent | Observable<NoPermissionEvent> | Observable<HasPermissionEvent>>' is not assignable to type 'Subscribable<EventObject>'.
            Types of property 'subscribe' are incompatible.
```